internal/server/data/migrator/helpers.go

```diff
@@ -55,3 +55,30 @@ func HasColumn(tx *gorm.DB, table string, column string) bool {
 	}
 	return count != 0
 }
+
+// HasConstraint returns true if the database table has the constraint. Returns
+// false if the database table does not have the constraint, or if there was a
+// failure querying the database.
+func HasConstraint(tx *gorm.DB, table string, constraint string) bool {
+	var count int
+	stmt := `
+		SELECT count(*)
+		FROM information_schema.table_constraints
+		WHERE table_schema = CURRENT_SCHEMA()
+		AND table_name = ? AND constraint_name = ?
+	`
+	if tx.Dialector.Name() == "sqlite" {
+		stmt = `
+			SELECT count(*)
+			FROM sqlite_master
+			WHERE type = 'table' AND tbl_name = ?
+			AND sql LIKE ?
+		`
+		constraint = "%CONSTRAINT `" + constraint + "`%"
+	}
+
+	if err := tx.Raw(stmt, table, constraint).Scan(&count).Error; err != nil {
+		return false
```

Comment:
Is there a reason to suppress this error rather than fail?

Reply:
Good question! My reasoning was:

1. this is what gorm did, so matching that behaviour should be fine if it was working previously
2. the only way this could fail in production would be a runtime error because the database is unavailable. If the database is unavailable then it will return false, which will cause the migration to run, and that error should be the same connection error as we would have seen here.

I think that second statement would be a lot safer if we had tests specifically for these function. I can add those.